### DISCUSSION TITLE
프로세스와 스레드의 차이(Process vs Thread)를 알려주세요.

### DISCUSSION THREAD

**Discussion by eatingbug** [Team-Psyduck/tech-sharing#62] [General]
1. 멀티프로세싱 시스템과 멀티스레딩 시스템 각각의 장점과 단점은 무엇인가요? 특정 상황에서는 왜 하나를 선택해야 하는지 설명해 주실 수 있나요?
2. 자식 프로세스가 부모 프로세스로부터 어떤 자원을 공유하나요? 반면에 스레드는 어떤 자원을 공유하나요?
3. 스레드와 프로세스 간의 컨텍스트 스위칭(Context Switching)에 대해 설명해 주세요. 이 둘 사이에서 컨텍스트 스위칭이 어떻게 다르며, 그 차이가 시스템의 성능에 어떤 영향을 미치는지 설명해 주세요.

**Comment by kkg5** [comment]
프로세스는 실행되고 있는 프로그램을 말하고 스레드는 프로세스 내의 가장 작은 작업 단위입니다.

**Comment by SionBackEnd** [comment]
프로세스와 스레드의 차이점으로 메모리 주소를 공유하냐 안 하냐로 볼 수 있습니다. 
프로세스는 code, data, stack영역을 공유하지 않지만, 스레드는 code와 data 영역은 공유하며 stack영역만을 공유하지 않습니다.(PC, register 공유 x)
이러한 차이점은 컨텍스트 스위치작업을 할 때 분명한 차이점이 발생합니다. 동이한 프로세스 안에서 스레드들끼리의 스위치 작업이 프로세스의 스위치 작업보다 더욱 빠릅니다. 스레드가 더 빠른 이유는 공유하고 있는 영역에 대한 스위치 작업을 할 필요가 없어서 더욱 빠르며 프로세스의 스위치 작업은 pc, register뿐 아니라 CPU 캐시를 완전히 비워주는 작업까지 진행해야 하기 때문에 더욱 느리다고 말할 수 있습니다. 

멀티프로세싱 시스템, 멀티 스레딩 장단점
멀티프로세싱 시스템은 안정성과 확장성을 높일 수 있지만, 오버헤드가 발생하고 프로세스 간 공유 데이터 처리를 위해 별도의 메커니즘이 필요합니다. 멀티스레딩 시스템은 빠른 응답 시간, 자원 공유 및 경제성의 장점을 가지고 있으나, 동기화 문제나 Deadlock(교착상태) 문제 같은 복잡한 동시성 문제들이 발생할 수 있습니다.

자식 프로세스는 부모 프로세스로부터 모든 자원을 물려받게 됩니다. 심지어 pc, register까지 물려받게 되어서 둘은 완전히 동일한 프로세스라고 말할 수 있습니다.

**Reply by eatingbug** [reply]
1. 멀티스레딩 시스템에서 동기화 문제나 교착 상태 문제를 방지하거나 해결하는 일반적인 방법은 무엇인가요?
2. 부모 프로세스와 자식 프로세스가 "완전히 동일한 프로세스"라고 말하셨는데, 그렇다면 부모 프로세스와 자식 프로세스 사이에 차이점은 없는 것인가요? 있다면 어떤 것들인가요?

**Reply by SionBackEnd** [reply]
1. 동기화 문제를 해결하는 일반적인 방법으로 락을 이용하여 방법이 있습니다. 락의 종류로는 스핀락, 뮤텍스, 세마포어가 존재합니다. 
하지만, 락을 이용하게 되면서 데드락이 발생 할 수 있게 되는데 이때는 데드락 발생 원인 4가지 중 한가지를 성립하지 못하게 하거나, 회피, 회복, 무시 등의 방법이 있습니다. 
2. 부모 프로세스가 자식 프로세스를 fork()함수를 통해 생성하게 되면 완전히 동일한 프로세스로 생성하게 됩니다. PC와 register값 또한 똑같기 때문에 자식 프로세스는 부모 프로세스가 fork()함수를 호출한 이후 로직부터 실행하게 됩니다. 유일하게 다른 점이라면, PID값을 다르게 얻게 되는데, 부모 프로세스는 자식 프로세스의 PID값을 얻게 되고, 자식프로세스는 PID값으로 '0'을 얻게 됩니다.

**Comment by eatingbug** [comment]
- 프로세스란 운영체제로부터 시스템 자원을 할당받는 작업의 단위로 메모리에 올라와 실행되고 있는 프로그램의 인스턴스(독립적인 개체)를 의미한다.
  - 프로세스당 최소 1개의 스레드를 갖고 있다.
  - 각 프로세스는 별도의 주소 공간에서 실행되며, 한 프로세스는 다른 프로세스의 변수나 자료구조에 접근할 수 없다.
- 스레드란 프로세스 내에서 실행되는 흐름의 단위를 말한다. 
  - 프로세스 내에서 각각 Stack 영역만 따로 할당받고 Code, Data, Heap 영역은 공유한다.
  - 프로세스 내의 주소 공간이나 자원들을 같은 프로세스 내에 스레드끼리 공유하면서 실행된다.
  - 각각의 스레드는 별도의 레지스터와 스택을 갖고 있지만, 힙 메모리는 서로 읽고 쓸 수 있다.

[[OS] 프로세스와 스레드의 차이 - Heee's Development Blog](https://gmlwjd9405.github.io/2018/09/14/process-vs-thread.html)
